### DISCUSSION TITLE
please make the label name dynamic, not hardcoded

### DISCUSSION THREAD

**Discussion by andytanoko** [trallnag/prometheus-fastapi-instrumentator#256] [Q&A]
prometheus-fastapi-instrumentator is a good library, it can automatically generate metrics. but label name is hardcoded. is there any development plan to make it dynamic? following is the code snippet from metrics.py

    if should_include_handler:
        label_names.append("handler")
        info_attribute_names.append("modified_handler")

    if should_include_method:
        label_names.append("method")
        info_attribute_names.append("method")

    if should_include_status:
        label_names.append("status")
        info_attribute_names.append("modified_status")

**Comment by trallnag** [comment]
Hi, you can add a feature request, but it is currently not planned. This would blow up the interface even more. Have you considered writing your own instrumentation function that mimics the default metrics? Here is one example of this:

https://github.com/trallnag/prometheus-fastapi-instrumentator/tree/master/devel/examples/default-metrics-diff-labels

**Reply by andytanoko** [reply]
this is the same as writing our own custom metrics.

**Reply by trallnag** [reply]
True. Another alternative is relabeling: https://stackoverflow.com/a/58065347/7391331

**Comment by alexisthethe** [comment]
I agree with @andytanoko ! That would be great if the labels could easily be customisable without having to rewrite the metrics functions.
In my case, I would love to be able to classify these library metrics "latency" and "requests" by user. In this case I would need to define a label that would get the info.request.headers.get("Authorization") and the function to extract the username from it.
So far, as the library is implemented, it is too difficult use inheritance to do it and I have to rewrite the library functions themselves.

I think we could change the way to get the label value.
Currently it is `getattr(info, attribute_name)` ([see code](https://github.com/trallnag/prometheus-fastapi-instrumentator/blob/b645ccb618ce9cbe8fd935868f359ebbc12ea217/src/prometheus_fastapi_instrumentator/metrics.py#L189C50-L189C50)).
Instead we could define a function in the class `Info` to get the label value and allow to add custom functions to get labels as a parameter in metrics functions.

I think this library is great, very useful. So I would be very happy to contribute and propose you a pull request for it. What do you think ?
